src/expr/src/relation/mod.rs

```diff
@@ -1129,6 +1174,27 @@ impl MirRelationExpr {
         self.visit1_mut(|e| e.visit_mut_pre(f))
     }
 
+    /// A generalization of `visit`. The function `pre` runs on a
+    /// `MirRelationExpr` before it runs on any of the child `MirRelationExpr`s.
+    /// The function `post` runs on child `MirRelationExpr`s first before the
+    /// parent. Optionally, `pre` can return which child `MirRelationExpr`s, if
+    /// any, should be visited (default is to visit all children).
+    pub fn visit_pre_post<F1, F2>(&self, pre: &mut F1, post: &mut F2)
```

Comment:
Is it right that this is a non-mut version of `visit_mut_pre_post`? If so, might it make sense to co-locate them? If there is already a good organizational principle in the file other than this, no worries!

Reply:
There is no mutable version of this method yet. `visit_mut_pre_post` exists only for `HirScalarExpr`, which in turn, doesn't have a non mutable one.

Reply:
I'm looking at `expr/src/scalar/mod.rs`, and there is a 
```rust
    /// A generalization of `visit_mut`. The function `pre` runs on a
    /// `MirScalarExpr` before it runs on any of the child `MirScalarExpr`s.
    /// The function `post` runs on child `MirScalarExpr`s first before the
    /// parent. Optionally, `pre` can return which child `MirScalarExpr`s, if
    /// any, should be visited (default is to visit all children).
    pub fn visit_mut_pre_post<F1, F2>(&mut self, pre: &mut F1, post: &mut F2)
    where
        F1: FnMut(&mut Self) -> Option<Vec<&mut MirScalarExpr>>,
        F2: FnMut(&mut Self),
    {
        let to_visit = pre(self);
        if let Some(to_visit) = to_visit {
            for e in to_visit {
                e.visit_mut_pre_post(pre, post);
            }
        } else {
            self.visit1_mut(|e| e.visit_mut_pre_post(pre, post));
        }
        post(self);
    }
```
in there. Am I confused?

Reply:
Yes, what I mean is that there is no `visit_mut_pre_post` in `MirRelationExpr`, only the new `visit_pre_post` I've added here.

Also, I was pointing out that in `MirScalarExpr` the opposite thing happens: it  has `visit_mut_pre_post`, but not `visit_pre_post`.

Reply:
I'm sorry, I read `HirScalarExpr` in your comment and thought that `Hir` was the point you were making. Instead you were making the point that it is `Scalar` rather than `Relation`.

Reply:
Yes, sorry. You read `HirScalarExpr` but I meant `MirScalarExpr`. I've fixed the typo.